controller/src/main/java/io/pravega/controller/store/stream/ZKStream.java

```diff
@@ -504,137 +699,177 @@
     }
 
     @Override
-    CompletableFuture<Data> getCommitTxnRecord() {
-        return store.getData(committingTxnsPath);
+    CompletableFuture<VersionedMetadata<CommittingTransactionsRecord>> getCommitTxnRecord() {
+        return store.getData(getCommittingTxnsPath(), CommittingTransactionsRecord::fromBytes);
     }
 
     @Override
-    CompletableFuture<Version> updateCommittingTxnRecord(Data update) {
-        return store.setData(committingTxnsPath, update)
+    CompletableFuture<Version> updateCommittingTxnRecord(VersionedMetadata<CommittingTransactionsRecord> update) {
+        return store.setData(getCommittingTxnsPath(), update.getObject().toBytes(), update.getVersion())
                     .thenApply(Version.IntVersion::new);
     }
 
     @Override
-    CompletableFuture<Void> createWaitingRequestNodeIfAbsent(byte[] waitingRequestProcessor) {
-        return Futures.toVoid(store.createZNodeIfNotExist(waitingRequestProcessorPath, waitingRequestProcessor));
+    CompletableFuture<Void> createWaitingRequestNodeIfAbsent(String waitingRequestProcessor) {
+        return Futures.toVoid(store.createZNodeIfNotExist(getWaitingRequestProcessorPath(), 
+                waitingRequestProcessor.getBytes(StandardCharsets.UTF_8)));
     }
 
     @Override
-    CompletableFuture<Data> getWaitingRequestNode() {
-        return store.getData(waitingRequestProcessorPath);
+    CompletableFuture<String> getWaitingRequestNode() {
+        return store.getData(getWaitingRequestProcessorPath(), x -> StandardCharsets.UTF_8.decode(ByteBuffer.wrap(x)).toString())
+                .thenApply(VersionedMetadata::getObject);
     }
 
     @Override
     CompletableFuture<Void> deleteWaitingRequestNode() {
-        return store.deletePath(waitingRequestProcessorPath, false);
+        return store.deletePath(getWaitingRequestProcessorPath(), false);
     }
 
     @Override
     public void refresh() {
-        // refresh all mutable records
-        cache.invalidateCache(statePath);
-        cache.invalidateCache(configurationPath);
-        cache.invalidateCache(truncationPath);
-        cache.invalidateCache(epochRecordPathFormat);
-        cache.invalidateCache(committingTxnsPath);
-        cache.invalidateCache(currentEpochRecordPath);
-        cache.invalidateCache(currentEpochRecordPath);
+        String id = this.idRef.getAndSet(null);
+        id = id == null ? "" : id;
+        // invalidate all mutable records in the cache 
+        store.invalidateCache(getStatePath(), id);
+        store.invalidateCache(getConfigurationPath(), id);
+        store.invalidateCache(getTruncationPath(), id);
+        store.invalidateCache(getEpochTransitionPath(), id);
+        store.invalidateCache(getCommittingTxnsPath(), id);
+        store.invalidateCache(getCurrentEpochRecordPath(), id);
+    }
+
+    /**
+     * Method to retrieve unique Id for the stream. We use streamPosition as the unique id for the stream.
```

Comment:
I don't understand this observation, how can we use `streamPosition` as an id? Specifically, I'm wondering about why it is correct to make an id out of a position.

Reply:
`... value, else`

Reply:
position is unique for each stream within a scope..
 making position as id makes it safe against recreated streams and cached values (as explained in the comment above)

Reply:
Remind me the definition of stream position so that I have more context for this, please. I still don't understand how it can be unique, but it is probably because I don't have enough context.

Reply:
Stream position is the unique number given to a stream at the time of its creation under the "StreamsInScope" tree. Everytime a stream gets created it gets a new higher number.